File: backend/app/api/documents.py
"""
Documents + Collections API router.

All endpoints here are:
  1. Authentication-gated (get_current_user is mandatory)
  2. Permission-gated via require_permission() dependency
  3. Org-scoped from the verified JWT (never from client-supplied params)

Endpoints:

  Documents:
    POST   /documents                         — upload (new doc or new version)
    GET    /documents                         — list with filters / pagination
    POST   /documents/bulk                    — bulk action (must come before /{id})
    GET    /documents/processing              — org-wide active jobs (Phase 4)
    GET    /documents/{id}                    — document detail
    PATCH  /documents/{id}                    — metadata update
    DELETE /documents/{id}                    — soft delete
    POST   /documents/{id}/restore            — restore soft-deleted doc
    GET    /documents/{id}/versions           — version history
    GET    /documents/{id}/download           — issue signed URL
    GET    /documents/{id}/status             — processing status (polling)
    GET    /documents/{id}/pages              — extracted pages text/OCR (Phase 5)
    GET    /documents/{id}/toc                — section tree for the TOC panel (Phase 6)
    GET    /documents/{id}/chunks             — chunks with provenance (debug, Phase 6)
    POST   /documents/{id}/retry              — retry failed stage (Phase 4)

  Collections:
    GET    /collections                       — list collections
    POST   /collections                       — create collection
    DELETE /collections/{id}                  — delete collection
    POST   /collections/{id}/documents        — add document to collection
    DELETE /collections/{id}/documents/{did}  — remove document from collection

See:
  Backend-Architecture-Documentation.md §17 (document flow)
  Database-Architecture-Design-Documentation.md §13–14, §19
"""
from __future__ import annotations

import logging
from datetime import datetime
from typing import Optional
from uuid import UUID

from fastapi import (
    APIRouter,
    Depends,
    Form,
    Query,
    Request,
    UploadFile,
    status,
)

from app.api.deps import DbSession, require_permission
from app.models.user import User
from app.schemas.document import (
    BulkActionRequest,
    BulkActionResponse,
    CollectionCreate,
    CollectionDocumentRequest,
    CollectionListResponse,
    CollectionResponse,
    DocumentChunksResponse,
    DocumentListResponse,
    DocumentMetadataUpdate,
    DocumentPagesResponse,
    DocumentResponse,
    DocumentStatusResponse,
    DocumentTocResponse,
    DocumentUploadResponse,
    DocumentVersionDetail,
    DownloadUrlResponse,
)
from app.schemas.processing import ProcessingListResponse, DocumentRetryResponse
from app.services.collection_service import CollectionService
from app.services.document_service import DocumentService
from app.services.job_service import JobService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/documents", tags=["documents"])
collections_router = APIRouter(prefix="/collections", tags=["collections"])


# ─── Upload ───────────────────────────────────────────────────────────────────

@router.post(
    "",
    response_model=DocumentUploadResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Upload a document (or a new version of an existing one)",
)
async def upload_document(
    request: Request,
    db: DbSession,
    file: UploadFile,
    name: str = Form(..., description="Document display name"),
    document_type: str = Form(...),
    department: Optional[str] = Form(None),
    description: Optional[str] = Form(None),
    access_level: str = Form("organization"),
    version_label: Optional[str] = Form(None),
    effective_date: Optional[str] = Form(None, description="ISO date (YYYY-MM-DD)"),
    document_id: Optional[str] = Form(None, alias="documentId"),
    user: User = Depends(require_permission("document:create")),
) -> DocumentUploadResponse:
    """Upload a PDF or DOCX.

    - Omit `documentId` to create a new logical document.
    - Supply `documentId` to add a new version to an existing document.

    Returns `202 Accepted` immediately. Processing (extraction, OCR, chunking,
    embeddings) happens asynchronously via the job queue (Phase 4+).
    """
    return await DocumentService.upload_document(
        file=file,
        organization_id=user.organization_id,
        owner_id=user.id,
        name=name,
        document_type=document_type,
        department=department,
        description=description,
        access_level=access_level,
        version_label=version_label,
        effective_date=effective_date,
        existing_document_id=document_id,
        db=db,
        request=request,
    )


# ─── List ─────────────────────────────────────────────────────────────────────

@router.get(
    "",
    response_model=DocumentListResponse,
    summary="List documents with optional filters and pagination",
)
async def list_documents(
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
    document_type: Optional[str] = Query(None, alias="type"),
    status: Optional[str] = Query(None),
    access_level: Optional[str] = Query(None, alias="accessLevel"),
    owner_id: Optional[str] = Query(None, alias="ownerId"),
    collection_id: Optional[str] = Query(None, alias="collectionId"),
    department: Optional[str] = Query(None),
    search: Optional[str] = Query(None, description="Name search (case-insensitive)"),
    limit: int = Query(20, ge=1, le=100),
    cursor_created_at: Optional[datetime] = Query(None, alias="cursorCreatedAt"),
    cursor_id: Optional[str] = Query(None, alias="cursorId"),
    sort_desc: bool = Query(True, alias="sortDesc"),
) -> DocumentListResponse:
    return await DocumentService.list_documents(
        organization_id=user.organization_id,
        document_type=document_type,
        status=status,
        access_level=access_level,
        owner_id=owner_id,
        collection_id=collection_id,
        department=department,
        search_name=search,
        cursor_created_at=cursor_created_at,
        cursor_id=cursor_id,
        limit=limit,
        sort_desc=sort_desc,
        db=db,
    )


# ─── Bulk action ──────────────────────────────────────────────────────────────
# IMPORTANT: /bulk must come BEFORE /{document_id} so FastAPI does not
# try to interpret the literal string "bulk" as a UUID path parameter.

@router.post(
    "/bulk",
    response_model=BulkActionResponse,
    summary="Apply an action to multiple documents",
)
async def bulk_action(
    payload: BulkActionRequest,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> BulkActionResponse:
    return await DocumentService.bulk_action(
        action=payload.action,
        document_ids=payload.document_ids,
        organization_id=user.organization_id,
        user_id=user.id,
        tags=payload.tags,
        access_level=payload.access_level,
        db=db,
        request=request,
    )


# ─── Org-wide active processing jobs (Phase 4) ────────────────────────────────
# IMPORTANT: /processing must come BEFORE /{document_id} so FastAPI does not
# try to interpret the literal string "processing" as a UUID path parameter.

@router.get(
    "/processing",
    response_model=ProcessingListResponse,
    summary="Org-wide active processing jobs (header indicator widget)",
)
async def list_processing_jobs(
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
    limit: int = Query(100, ge=1, le=500),
) -> ProcessingListResponse:
    """All PENDING/PROCESSING/RETRYING jobs in the organization.

    Polled by the global header processing indicator (FE §5.2); SSE arrives
    in Phase 11.
    """
    return await JobService.list_org_processing(
        organization_id=user.organization_id,
        db=db,
        limit=limit,
    )


# ─── Detail ───────────────────────────────────────────────────────────────────

@router.get(
    "/{document_id}",
    response_model=DocumentResponse,
    summary="Get document detail",
)
async def get_document(
    document_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
) -> DocumentResponse:
    return await DocumentService.get_document(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


# ─── Metadata update ──────────────────────────────────────────────────────────

@router.patch(
    "/{document_id}",
    response_model=DocumentResponse,
    summary="Update document metadata",
)
async def update_document(
    document_id: str,
    payload: DocumentMetadataUpdate,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> DocumentResponse:
    return await DocumentService.update_metadata(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        update=payload,
        db=db,
        request=request,
    )


# ─── Soft delete ──────────────────────────────────────────────────────────────

@router.delete(
    "/{document_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    summary="Soft-delete a document",
)
async def delete_document(
    document_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:delete")),
) -> None:
    await DocumentService.soft_delete(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


# ─── Restore ──────────────────────────────────────────────────────────────────

@router.post(
    "/{document_id}/restore",
    response_model=DocumentResponse,
    summary="Restore a soft-deleted document",
)
async def restore_document(
    document_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> DocumentResponse:
    return await DocumentService.restore(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


# ─── Versions ─────────────────────────────────────────────────────────────────

@router.get(
    "/{document_id}/versions",
    response_model=list[DocumentVersionDetail],
    summary="List version history for a document",
)
async def get_versions(
    document_id: str,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
) -> list[DocumentVersionDetail]:
    return await DocumentService.get_version_list(
        document_id=document_id,
        organization_id=user.organization_id,
        db=db,
    )


# ─── Download ─────────────────────────────────────────────────────────────────

@router.get(
    "/{document_id}/download",
    response_model=DownloadUrlResponse,
    summary="Get a short-lived signed URL to download the document",
)
async def get_download_url(
    document_id: str,
    request: Request,
    db: DbSession,
    version: Optional[int] = Query(None, description="Version number (omit for latest)"),
    user: User = Depends(require_permission("document:read")),
) -> DownloadUrlResponse:
    return await DocumentService.get_download_url(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        version_number=version,
        db=db,
        request=request,
    )


# ─── Status ───────────────────────────────────────────────────────────────────

@router.get(
    "/{document_id}/status",
    response_model=DocumentStatusResponse,
    summary="Get processing status for the latest version",
)
async def get_document_status(
    document_id: str,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
) -> DocumentStatusResponse:
    """Polling snapshot: `{ status, currentStep, progress, errorMessage? }` plus
    live job detail. SSE upgrade arrives in Phase 11 (FE §11)."""
    return await JobService.get_document_processing_status(
        document_id=document_id,
        organization_id=user.organization_id,
        db=db,
    )


# ─── Extracted pages (Phase 5) ────────────────────────────────────────────────

@router.get(
    "/{document_id}/pages",
    response_model=DocumentPagesResponse,
    summary="Extracted page text + OCR flags for one version",
)
async def get_document_pages(
    document_id: str,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
    version: Optional[int] = Query(None, description="Version number (omit for latest)"),
    offset: int = Query(0, ge=0, alias="offset"),
    limit: int = Query(100, ge=1, le=500),
) -> DocumentPagesResponse:
    """Page-level extraction results (workspace/debug tooling).

    Returns pages persisted so far — safe to read mid-extraction (shows
    partial progress). `ocrFailed` per page drives the partial-processing
    warning in the workspace.
    """
    return await DocumentService.get_document_pages(
        document_id=document_id,
        organization_id=user.organization_id,
        version_number=version,
        offset=offset,
        limit=limit,
        db=db,
    )


# ─── Table of contents (Phase 6) ──────────────────────────────────────────────

@router.get(
    "/{document_id}/toc",
    response_model=DocumentTocResponse,
    summary="Section tree for one version (workspace TOC panel)",
)
async def get_document_toc(
    document_id: str,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
    version: Optional[int] = Query(None, description="Version number (omit for latest)"),
) -> DocumentTocResponse:
    """Hierarchical table of contents detected during chunking (Phase 6).

    An empty `items` list is the explicit "No structure detected" state
    (FE §6.5) — the workspace falls back to page-based navigation.
    """
    return await DocumentService.get_document_toc(
        document_id=document_id,
        organization_id=user.organization_id,
        version_number=version,
        db=db,
    )


# ─── Chunks (Phase 6 — internal chunk-debug tooling) ──────────────────────────

@router.get(
    "/{document_id}/chunks",
    response_model=DocumentChunksResponse,
    summary="Chunks of one version with provenance (debug tooling)",
)
async def get_document_chunks(
    document_id: UUID,
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
    version: Optional[int] = Query(None, description="Version number (omit for latest)"),
    offset: int = Query(0, ge=0, alias="offset"),
    limit: int = Query(50, ge=1, le=500),
) -> DocumentChunksResponse:
    """The retrieval units produced by the CHUNKING stage, with full
    provenance (section, heading path, pages, flags) — the chunk-quality
    debugging view that later phases (embeddings/search/citations) build on.
    """
    return await DocumentService.get_document_chunks(
        document_id=str(document_id),
        organization_id=user.organization_id,
        version_number=version,
        offset=offset,
        limit=limit,
        db=db,
    )


# ─── Retry (Phase 4 — explicit FAILED → PROCESSING action) ────────────────────

@router.post(
    "/{document_id}/retry",
    response_model=DocumentRetryResponse,
    status_code=status.HTTP_202_ACCEPTED,
    summary="Retry a failed processing job (explicit action)",
)
async def retry_processing(
    document_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> DocumentRetryResponse:
    """Re-create the failed stage's job and transition FAILED → PROCESSING.

    Retry is NEVER automatic (Backend §47) — this is the deliberate user
    action. Raises 409 when there is no failed job or processing is already
    in flight.
    """
    return await JobService.retry_failed_stage(
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


# ─── Collections router ───────────────────────────────────────────────────────

@collections_router.get(
    "",
    response_model=CollectionListResponse,
    summary="List all collections in the organization",
)
async def list_collections(
    db: DbSession,
    user: User = Depends(require_permission("document:read")),
) -> CollectionListResponse:
    return await CollectionService.list_collections(
        organization_id=user.organization_id,
        db=db,
    )


@collections_router.post(
    "",
    response_model=CollectionResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Create a new collection",
)
async def create_collection(
    payload: CollectionCreate,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:create")),
) -> CollectionResponse:
    return await CollectionService.create_collection(
        organization_id=user.organization_id,
        user_id=user.id,
        payload=payload,
        db=db,
        request=request,
    )


@collections_router.delete(
    "/{collection_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    summary="Delete a collection (does not delete the documents)",
)
async def delete_collection(
    collection_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:delete")),
) -> None:
    await CollectionService.delete_collection(
        collection_id=collection_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


@collections_router.post(
    "/{collection_id}/documents",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    summary="Add a document to a collection",
)
async def add_document_to_collection(
    collection_id: str,
    payload: CollectionDocumentRequest,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> None:
    await CollectionService.add_document(
        collection_id=collection_id,
        document_id=payload.document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )


@collections_router.delete(
    "/{collection_id}/documents/{document_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    response_model=None,
    summary="Remove a document from a collection",
)
async def remove_document_from_collection(
    collection_id: str,
    document_id: str,
    request: Request,
    db: DbSession,
    user: User = Depends(require_permission("document:update")),
) -> None:
    await CollectionService.remove_document(
        collection_id=collection_id,
        document_id=document_id,
        organization_id=user.organization_id,
        user_id=user.id,
        db=db,
        request=request,
    )

File: frontend/src/lib/api/documents.ts
/**
 * Typed API functions for documents + background processing (Phase 4).
 *
 * Matches backend/app/schemas/document.py and backend/app/schemas/processing.py.
 * Processing-status endpoints are the Phase 4 polling transport; the SSE
 * upgrade (Phase 11) will consume the same response shapes.
 */

import { del, get, post } from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

/** Pipeline status of a document version (document_versions.status). */
export type VersionStatus =
  | 'UPLOADED'
  | 'PROCESSING'
  | 'EXTRACTING'
  | 'OCR'
  | 'CHUNKING'
  | 'EMBEDDING'
  | 'INDEXING'
  | 'READY'
  | 'FAILED'

export type JobStatus = 'PENDING' | 'PROCESSING' | 'RETRYING' | 'COMPLETED' | 'FAILED'

export type JobType =
  | 'EXTRACTION'
  | 'OCR'
  | 'CHUNKING'
  | 'EMBEDDING'
  | 'INDEXING'
  | 'COMPARISON'
  | 'SUMMARY'
  | 'CONFLICT_SCAN'
  | 'PURGE'

/** GET /documents/{id}/status — processing status snapshot (polling). */
export interface DocumentStatusResponse {
  document_id: string
  version_id: string
  version_number: number
  status: VersionStatus
  /** 0–100, populated by workers while a stage is running */
  progress: number | null
  /** e.g. "340/512 chunks embedded" */
  progress_message: string | null
  error_message: string | null
  /** job_type of the in-flight stage, e.g. EXTRACTION */
  current_step: JobType | null
  job_id: string | null
  job_status: JobStatus | null
}

/** One active job — item of GET /documents/processing. */
export interface ProcessingJobItem {
  job_id: string
  document_id: string
  document_name: string
  version_id: string
  version_number: number
  job_type: JobType
  status: JobStatus
  progress: number | null
  progress_message: string | null
  attempts: number
  max_attempts: number
  started_at: string | null
  created_at: string
}

/** GET /documents/processing — org-wide active jobs (header widget). */
export interface ProcessingListResponse {
  items: ProcessingJobItem[]
  total: number
}

/** POST /documents/{id}/retry — 202 Accepted. */
export interface DocumentRetryResponse {
  job_id: string
  job_type: JobType
  status: JobStatus
  version_status: VersionStatus
  message: string
}

// ─── Document detail + pages (Phase 5) ────────────────────────────────────────

/** Version summary embedded in document detail responses. */
export interface DocumentVersionSummary {
  id: string
  version_number: number
  version_label: string | null
  status: VersionStatus
  mime_type: string
  file_size_bytes: number
  page_count: number | null
  effective_date: string | null
  expiration_date: string | null
  created_at: string
  created_by: string
  error_message: string | null
}

/** GET /documents/{id} — document detail. */
export interface DocumentDetail {
  id: string
  organization_id: string
  name: string
  description: string | null
  document_type: string
  department: string | null
  status: string
  access_level: string
  owner_id: string
  current_version_id: string | null
  created_at: string
  updated_at: string
  deleted_at: string | null
  tags: string[]
  current_version: DocumentVersionSummary | null
  version_count: number
}

/** GET /documents/{id}/download — short-lived signed URL. */
export interface DownloadUrlResponse {
  url: string
  expires_at: string
  mime_type: string
  file_size_bytes: number
  version_number: number
}

/** Lightweight document — item of GET /documents (list views + pickers). */
export interface DocumentListItem {
  id: string
  name: string
  document_type: string
  department: string | null
  status: string
  access_level: string
  owner_id: string
  current_version_id: string | null
  created_at: string
  updated_at: string
  tags: string[]
  processing_status: string | null
  page_count: number | null
}

/** GET /documents — document list response envelope. */
export interface DocumentListResponse {
  items: DocumentListItem[]
  total: number
}

/** One extracted page — item of GET /documents/{id}/pages. */
export interface DocumentPageItem {
  page_number: number
  text: string
  ocr_used: boolean
  ocr_failed: boolean
  width: number | null
  height: number | null
}

/** GET /documents/{id}/pages — extracted pages of one version (Phase 5). */
export interface DocumentPagesResponse {
  document_id: string
  version_id: string
  version_number: number
  mime_type: string
  status: VersionStatus
  page_count: number | null
  total: number
  items: DocumentPageItem[]
}

/** POST /documents (multipart upload) — 202 Accepted. */
export interface DocumentUploadResponse {
  document_id: string
  version_id: string
  version_number: number
  status: string
  is_duplicate_warning: boolean
  duplicate_version_id: string | null
  job_id: string | null
}

/** GET /documents/{id}/content — one page for the citation source panel. */
export interface DocumentPageContentResponse {
  document_id: string
  version_id: string
  version_number: number
  status: VersionStatus
  page_number: number
  page_count: number | null
  text: string
  ocr_used: boolean
  ocr_failed: boolean
  width: number | null
  height: number | null
}

// ─── Table of contents + chunks (Phase 6) ─────────────────────────────────────

/** One node of the section tree — GET /documents/{id}/toc. */
export interface TocNode {
  id: string
  title: string
  section_number: string | null
  level: number
  start_page: number
  end_page: number | null
  sort_order: number
  children: TocNode[]
}

/** GET /documents/{id}/toc — section tree for one version (Phase 6). */
export interface DocumentTocResponse {
  document_id: string
  version_id: string
  version_number: number
  /** false = "No structure detected" (FE §6.5) — page navigation only */
  has_structure: boolean
  section_count: number
  items: TocNode[]
}

/** One chunk with provenance — item of GET /documents/{id}/chunks. */
export interface DocumentChunkItem {
  chunk_index: number
  content: string
  token_count: number
  content_hash: string
  section_id: string | null
  section_number: string | null
  heading_path: string[]
  start_page: number | null
  end_page: number | null
  contains_table: boolean
  contains_list: boolean
  forced_split: boolean
  has_embedding: boolean
}

/** GET /documents/{id}/chunks — chunks of one version (debug tooling). */
export interface DocumentChunksResponse {
  document_id: string
  version_id: string
  version_number: number
  status: VersionStatus
  total: number
  items: DocumentChunkItem[]
}

// ─── Endpoints ────────────────────────────────────────────────────────────────

/** Processing status of a document's latest version (poll every few seconds). */
export function getDocumentStatusApi(documentId: string): Promise<DocumentStatusResponse> {
  return get<DocumentStatusResponse>(`/documents/${documentId}/status`)
}

/** Org-wide active jobs — drives the global header processing indicator. */
export function listProcessingJobsApi(limit = 100): Promise<ProcessingListResponse> {
  return get<ProcessingListResponse>(`/documents/processing?limit=${limit}`)
}

/**
 * Explicit retry of a failed stage (FAILED → PROCESSING).
 * Retries are NEVER automatic — this is the deliberate user action.
 */
export function retryProcessingApi(documentId: string): Promise<DocumentRetryResponse> {
  return post<DocumentRetryResponse>(`/documents/${documentId}/retry`)
}

/** Document detail (metadata + current version summary). */
export function getDocumentApi(documentId: string): Promise<DocumentDetail> {
  return get<DocumentDetail>(`/documents/${documentId}`)
}

/** Short-lived signed URL to the original file (never proxied by the API). */
export function getDownloadUrlApi(
  documentId: string,
  params: { version?: number } = {},
): Promise<DownloadUrlResponse> {
  const suffix = params.version != null ? `?version=${params.version}` : ''
  return get<DownloadUrlResponse>(`/documents/${documentId}/download${suffix}`)
}

/** Extracted page text + per-page OCR flags for a version (Phase 5). */
export function getDocumentPagesApi(
  documentId: string,
  params: { offset?: number; limit?: number; version?: number } = {},
): Promise<DocumentPagesResponse> {
  const query = new URLSearchParams()
  if (params.offset != null) query.set('offset', String(params.offset))
  if (params.limit != null) query.set('limit', String(params.limit))
  if (params.version != null) query.set('version', String(params.version))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return get<DocumentPagesResponse>(`/documents/${documentId}/pages${suffix}`)
}

/** Section tree for a version (Phase 6 — workspace TOC panel). */
export function getDocumentTocApi(
  documentId: string,
  version?: number,
): Promise<DocumentTocResponse> {
  const suffix = version != null ? `?version=${version}` : ''
  return get<DocumentTocResponse>(`/documents/${documentId}/toc${suffix}`)
}

/** Chunks with full provenance (Phase 6 — internal chunk-debug tooling). */
export function getDocumentChunksApi(
  documentId: string,
  params: { offset?: number; limit?: number } = {},
): Promise<DocumentChunksResponse> {
  const query = new URLSearchParams()
  if (params.offset != null) query.set('offset', String(params.offset))
  if (params.limit != null) query.set('limit', String(params.limit))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return get<DocumentChunksResponse>(`/documents/${documentId}/chunks${suffix}`)
}

/** Document list — powers library views and pickers (e.g. Ask AI scope, Phase 9). */
export function listDocumentsApi(params: { limit?: number; offset?: number } = {}): Promise<DocumentListResponse> {
  const query = new URLSearchParams()
  if (params.limit != null) query.set('limit', String(params.limit))
  if (params.offset != null) query.set('offset', String(params.offset))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return get<DocumentListResponse>(`/documents${suffix}`)
}

// ─── Library page additions (Phase 15) ────────────────────────────────────────

/** Filter/sort params for the Documents library page (backend list contract). */
export interface DocumentListParams {
  limit?: number
  offset?: number
  search?: string
  document_type?: string
  department?: string
  status?: 'active' | 'archived'
  /** Backend returns newest-first by default (sortDesc=true). */
  sortDesc?: boolean
}

/** GET /documents with the full filter set (Documents library, Phase 15). */
export function getDocumentListApi(params: DocumentListParams = {}): Promise<DocumentListResponse> {
  const query = new URLSearchParams()
  if (params.limit != null) query.set('limit', String(params.limit))
  if (params.offset != null) query.set('offset', String(params.offset))
  if (params.search) query.set('search', params.search)
  if (params.document_type) query.set('type', params.document_type)
  if (params.department) query.set('department', params.department)
  if (params.status) query.set('status', params.status)
  if (params.sortDesc != null) query.set('sortDesc', String(params.sortDesc))
  const suffix = query.toString() ? `?${query.toString()}` : ''
  return get<DocumentListResponse>(`/documents${suffix}`)
}

export interface UploadDocumentParams {
  file: File
  name: string
  document_type: string
  department?: string
  description?: string
  access_level?: 'organization' | 'restricted' | 'private'
  /** Supply to add a new version to an existing document. */
  document_id?: string
}

/** POST /documents multipart upload — 202 Accepted (processing is async). */
export function uploadDocumentApi(params: UploadDocumentParams): Promise<DocumentUploadResponse> {
  const form = new FormData()
  form.append('file', params.file)
  form.append('name', params.name)
  form.append('document_type', params.document_type)
  if (params.department) form.append('department', params.department)
  if (params.description) form.append('description', params.description)
  if (params.access_level) form.append('access_level', params.access_level)
  if (params.document_id) form.append('documentId', params.document_id)
  return post<DocumentUploadResponse>('/documents', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

/** DELETE /documents/{id} — soft delete (202 Accepted). */
export function deleteDocumentApi(documentId: string): Promise<void> {
  return del<void>(`/documents/${documentId}`)
}

/** GET /documents/{id}/content — one page's content (citation source panel). */
export function getDocumentContentApi(
  documentId: string,
  params: { page: number; version?: number },
): Promise<DocumentPageContentResponse> {
  const query = new URLSearchParams({ page: String(params.page) })
  if (params.version != null) query.set('version', String(params.version))
  return get<DocumentPageContentResponse>(
    `/documents/${documentId}/content?${query.toString()}`,
  )
}
